src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp

```diff
@@ -95,7 +904,1269 @@ class boss_thorim : public CreatureScript
+        {
+            if (SpellInfo const* spellInfo = sSpellMgr->GetSpellInfo(spellId))
+                return GetTotalHeal(spellInfo, caster);
+            return 0;
+        }
+
+        /// returns remaining heal amount on given target
+        static uint32 GetRemainingHealOn(Unit* target)
+        {
+            uint32 heal = 0;
+            Unit::AuraApplicationMap const& auras = target->GetAppliedAuras();
+            for (Unit::AuraApplicationMap::const_iterator itr = auras.begin(); itr != auras.end(); ++itr)
+            {
+                Aura const* aura = itr->second->GetBase();
+
+                for (uint8 i = 0; i < MAX_SPELL_EFFECTS; ++i)
+                {
+                    if (AuraEffect const* aurEff = aura->GetEffect(i))
+                    {
+                        if (aurEff->GetAuraType() == SPELL_AURA_PERIODIC_HEAL)
+                            heal += aurEff->GetAmount() * (aurEff->GetTotalTicks() - aurEff->GetTickNumber());
+                    }
+                }
+            }
+            return heal;
+        }
+
+        class MostHPMissingInRange
+        {
+            public:
+                MostHPMissingInRange(Unit const* referer, float range, uint32 hp, uint32 exclAura = 0, bool exclSelf = false)
+                    : _referer(referer), _range(range), _hp(hp), _exclAura(exclAura), _exclSelf(exclSelf) { }
+
+                bool operator()(Unit* u)
+                {
+                    if (_exclSelf && u == _referer)
+                        return false;
+
+                    if (_exclAura && u->HasAura(sSpellMgr->GetSpellIdForDifficulty(_exclAura, _referer)))
+                        return false;
+
+                    if ((u->GetHealth() + GetRemainingHealOn(u) + _hp) > u->GetMaxHealth())
+                        return false;
+
+                    uint32 missingHP = u->GetMaxHealth() - u->GetHealth();
+                    if (u->IsAlive() && _referer->IsFriendlyTo(u) && _referer->IsWithinDistInMap(u, _range) && missingHP > _hp)
+                    {
+                        _hp = missingHP;
+                        return true;
+                    }
+
+                    return false;
+                }
+
+            private:
+                Unit const* _referer;
+                float _range;
+                uint32 _hp;
+                uint32 _exclAura;
+                bool _exclSelf;
+        };
+
+        static Unit* GetUnitWithMostMissingHp(SpellInfo const* spellInfo, Unit* caster)
+        {
+            float range = spellInfo->GetMaxRange(false);
+            uint32 heal = GetTotalHeal(spellInfo, caster);
+
+            Unit* target = nullptr;
+            MostHPMissingInRange checker(caster, range, heal);
+            Trinity::UnitLastSearcher<MostHPMissingInRange> searcher(caster, target, checker);
+            caster->VisitNearbyObject(range, searcher);
+
+            return target;
+        }
+
+        static Unit* GetHealTarget(uint32 spellId, Unit* caster)
```

Comment:
There's a coupling in that case. Searchers are just gimme any target in range. AI then decides what to do with those targets.

IMO Searchers are good as they are